File: sdk-javascript/src/exceptions/ErrorCode.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import { Tier } from './Tier.js';

/**
 * Single source of truth for every error code the SDK raises. Cross-SDK
 * invariant: the wire-format codes here match the Java SDK's `ErrorCode`
 * enum, the Python SDK's `hfcx_sdk.exceptions.ErrorCode`, and the .NET
 * SDK's `ErrorCode` byte-for-byte. 27 entries: 9 protocol, 12 business,
 * 6 technical.
 */
export interface ErrorCodeEntry {
  readonly code: string;
  readonly tier: Tier;
  readonly description: string;
}

const make = (code: string, tier: Tier, description: string): ErrorCodeEntry => ({
  code,
  tier,
  description,
});

export const ErrorCode = {
  // ── ERR-P-* — protocol / wire-format violations ────────────────────────
  MISSING_HEADER: make('ERR-P-001', Tier.PROTOCOL, 'Required protocol header missing or empty'),
  JWE_ALGORITHM_REJECTED: make(
    'ERR-P-002',
    Tier.PROTOCOL,
    'JWE algorithm pair is not the pinned RSA-OAEP-256 / A256GCM',
  ),
  RECIPIENT_CODE_MISMATCH: make(
    'ERR-P-003',
    Tier.PROTOCOL,
    'x-hcx-recipient_code does not match this participant',
  ),
  BAD_UUID: make('ERR-P-004', Tier.PROTOCOL, 'Header value is not a valid UUID'),
  BAD_TIMESTAMP: make(
    'ERR-P-005',
    Tier.PROTOCOL,
    'x-hcx-timestamp is not a valid ISO-8601 instant',
  ),
  TIMESTAMP_OUT_OF_RANGE: make(
    'ERR-P-006',
    Tier.PROTOCOL,
    'x-hcx-timestamp is outside the configured tolerance window',
  ),
  SENDER_UNKNOWN: make('ERR-P-007', Tier.PROTOCOL, 'Sender participant code is not registered'),
  BAD_ENVELOPE: make('ERR-P-008', Tier.PROTOCOL, 'Request body envelope is malformed'),
  SIGNATURE_VERIFICATION_FAILED: make(
    'ERR-P-009',
    Tier.PROTOCOL,
    'Detached signature verification failed',
  ),

  // ── ERR-B-* — business / FHIR / Egyptian profile violations ────────────
  PARTICIPANT_NOT_FOUND: make(
    'ERR-B-001',
    Tier.BUSINESS,
    'Participant code not found in the registry',
  ),
  NOT_A_BUNDLE: make('ERR-B-002', Tier.BUSINESS, 'Top-level FHIR resource is not a Bundle'),
  BUNDLE_MISSING_TYPE: make(
    'ERR-B-003',
    Tier.BUSINESS,
    'Bundle.type is required by the Egyptian IG',
  ),
  PATIENT_MISSING_NATIONAL_ID: make(
    'ERR-B-004',
    Tier.BUSINESS,
    'Patient resource missing the National-ID identifier slice',
  ),
  PATIENT_NON_EGYPTIAN: make('ERR-B-005', Tier.BUSINESS, "Patient.address[0].country must be 'EG'"),
  NATIONAL_ID_INVALID: make(
    'ERR-B-006',
    Tier.BUSINESS,
    'Egyptian National ID value fails structural validation',
  ),
  PHONE_INVALID: make(
    'ERR-B-007',
    Tier.BUSINESS,
    'Egyptian mobile phone value is not in a recognised format',
  ),
  IBAN_INVALID: make(
    'ERR-B-008',
    Tier.BUSINESS,
    'Egyptian IBAN value fails the ISO 13616 mod-97 check',
  ),
  BAD_FHIR_JSON: make('ERR-B-009', Tier.BUSINESS, 'FHIR payload is not valid JSON'),
  ENVELOPE_MISSING_PAYLOAD: make(
    'ERR-B-010',
    Tier.BUSINESS,
    "Request body envelope is missing the 'payload' field",
  ),
  ENVELOPE_MALFORMED_JSON: make('ERR-B-011', Tier.BUSINESS, 'Request body is not valid JSON'),
  UNKNOWN_BUSINESS: make(
    'ERR-B-012',
    Tier.BUSINESS,
    'Unspecified business-rule failure (gateway error code missing or unparseable)',
  ),

  // ── ERR-T-* — technical / transport failures ──────────────────────────
  TRANSPORT: make('ERR-T-001', Tier.TECHNICAL, 'Transport-layer failure'),
  AUTHENTICATION: make(
    'ERR-T-002',
    Tier.TECHNICAL,
    'Authentication rejected by the identity provider',
  ),
  REGISTRY_UNAVAILABLE: make('ERR-T-003', Tier.TECHNICAL, 'Participant registry is unreachable'),
  KEY_UNAVAILABLE: make('ERR-T-004', Tier.TECHNICAL, 'Recipient private key cannot be loaded'),
  CRYPTOGRAPHIC_FAILURE: make(
    'ERR-T-005',
    Tier.TECHNICAL,
    'JOSE library reported a cryptographic failure',
  ),
  GATEWAY_5XX: make(
    'ERR-T-006',
    Tier.TECHNICAL,
    'HFCX gateway returned a 5xx response after retry exhaustion',
  ),
} as const;

export type ErrorCodeName = keyof typeof ErrorCode;

/** Every defined error-code entry, in declaration order. */
export const ALL_ERROR_CODES: readonly ErrorCodeEntry[] = Object.values(ErrorCode);

const _byWireCode: ReadonlyMap<string, ErrorCodeEntry> = new Map(
  ALL_ERROR_CODES.map((entry) => [entry.code, entry]),
);

/** Look up an entry by its wire-format code, or `undefined` if not found. */
export function errorCodeFromWire(wireCode: string | null | undefined): ErrorCodeEntry | undefined {
  if (!wireCode) return undefined;
  return _byWireCode.get(wireCode);
}

File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Exceptions/ErrorCode.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace HealthFlow.Hfcx.Sdk.Exceptions;

/// <summary>
/// Single source of truth for every error code the SDK raises. Cross-SDK
/// invariant: the wire-format codes here match the Java SDK's
/// <c>ErrorCode</c> enum and the Python SDK's <c>hfcx_sdk.exceptions.ErrorCode</c>
/// byte-for-byte. 27 entries: 9 protocol, 12 business, 6 technical.
/// </summary>
public sealed class ErrorCode
{
    // ── ERR-P-* — protocol / wire-format violations ─────────────────────
    public static readonly ErrorCode MissingHeader = new("ERR-P-001", Tier.Protocol, "Required protocol header missing or empty");
    public static readonly ErrorCode JweAlgorithmRejected = new("ERR-P-002", Tier.Protocol, "JWE algorithm pair is not the pinned RSA-OAEP-256 / A256GCM");
    public static readonly ErrorCode RecipientCodeMismatch = new("ERR-P-003", Tier.Protocol, "x-hcx-recipient_code does not match this participant");
    public static readonly ErrorCode BadUuid = new("ERR-P-004", Tier.Protocol, "Header value is not a valid UUID");
    public static readonly ErrorCode BadTimestamp = new("ERR-P-005", Tier.Protocol, "x-hcx-timestamp is not a valid ISO-8601 instant");
    public static readonly ErrorCode TimestampOutOfRange = new("ERR-P-006", Tier.Protocol, "x-hcx-timestamp is outside the configured tolerance window");
    public static readonly ErrorCode SenderUnknown = new("ERR-P-007", Tier.Protocol, "Sender participant code is not registered");
    public static readonly ErrorCode BadEnvelope = new("ERR-P-008", Tier.Protocol, "Request body envelope is malformed");
    public static readonly ErrorCode SignatureVerificationFailed = new("ERR-P-009", Tier.Protocol, "Detached signature verification failed");

    // ── ERR-B-* — business / FHIR / Egyptian profile violations ─────────
    public static readonly ErrorCode ParticipantNotFound = new("ERR-B-001", Tier.Business, "Participant code not found in the registry");
    public static readonly ErrorCode NotABundle = new("ERR-B-002", Tier.Business, "Top-level FHIR resource is not a Bundle");
    public static readonly ErrorCode BundleMissingType = new("ERR-B-003", Tier.Business, "Bundle.type is required by the Egyptian IG");
    public static readonly ErrorCode PatientMissingNationalId = new("ERR-B-004", Tier.Business, "Patient resource missing the National-ID identifier slice");
    public static readonly ErrorCode PatientNonEgyptian = new("ERR-B-005", Tier.Business, "Patient.address[0].country must be 'EG'");
    public static readonly ErrorCode NationalIdInvalid = new("ERR-B-006", Tier.Business, "Egyptian National ID value fails structural validation");
    public static readonly ErrorCode PhoneInvalid = new("ERR-B-007", Tier.Business, "Egyptian mobile phone value is not in a recognised format");
    public static readonly ErrorCode IbanInvalid = new("ERR-B-008", Tier.Business, "Egyptian IBAN value fails the ISO 13616 mod-97 check");
    public static readonly ErrorCode BadFhirJson = new("ERR-B-009", Tier.Business, "FHIR payload is not valid JSON");
    public static readonly ErrorCode EnvelopeMissingPayload = new("ERR-B-010", Tier.Business, "Request body envelope is missing the 'payload' field");
    public static readonly ErrorCode EnvelopeMalformedJson = new("ERR-B-011", Tier.Business, "Request body is not valid JSON");
    public static readonly ErrorCode UnknownBusiness = new("ERR-B-012", Tier.Business, "Unspecified business-rule failure (gateway error code missing or unparseable)");

    // ── ERR-T-* — technical / transport failures ────────────────────────
    public static readonly ErrorCode Transport = new("ERR-T-001", Tier.Technical, "Transport-layer failure");
    public static readonly ErrorCode Authentication = new("ERR-T-002", Tier.Technical, "Authentication rejected by the identity provider");
    public static readonly ErrorCode RegistryUnavailable = new("ERR-T-003", Tier.Technical, "Participant registry is unreachable");
    public static readonly ErrorCode KeyUnavailable = new("ERR-T-004", Tier.Technical, "Recipient private key cannot be loaded");
    public static readonly ErrorCode CryptographicFailure = new("ERR-T-005", Tier.Technical, "JOSE library reported a cryptographic failure");
    public static readonly ErrorCode Gateway5xx = new("ERR-T-006", Tier.Technical, "HFCX gateway returned a 5xx response after retry exhaustion");

    private static readonly IReadOnlyList<ErrorCode> _all =
    [
        MissingHeader, JweAlgorithmRejected, RecipientCodeMismatch, BadUuid,
        BadTimestamp, TimestampOutOfRange, SenderUnknown, BadEnvelope,
        SignatureVerificationFailed,
        ParticipantNotFound, NotABundle, BundleMissingType, PatientMissingNationalId,
        PatientNonEgyptian, NationalIdInvalid, PhoneInvalid, IbanInvalid,
        BadFhirJson, EnvelopeMissingPayload, EnvelopeMalformedJson, UnknownBusiness,
        Transport, Authentication, RegistryUnavailable, KeyUnavailable,
        CryptographicFailure, Gateway5xx,
    ];

    private static readonly ReadOnlyDictionary<string, ErrorCode> _byCode = BuildIndex();

    /// <summary>Every defined error code, in declaration order.</summary>
    public static IReadOnlyList<ErrorCode> All => _all;

    /// <summary>Wire-format code, e.g. <c>"ERR-P-001"</c>.</summary>
    public string Code { get; }

    /// <summary>Tier this code belongs to.</summary>
    public Tier Tier { get; }

    /// <summary>Human-readable description.</summary>
    public string Description { get; }

    private ErrorCode(string code, Tier tier, string description)
    {
        Code = code;
        Tier = tier;
        Description = description;
    }

    /// <summary>
    /// Look up an <see cref="ErrorCode"/> by its wire-format code, or
    /// <see langword="null"/> if no entry matches.
    /// </summary>
    public static ErrorCode? FromWire(string? wireCode)
    {
        if (string.IsNullOrEmpty(wireCode))
        {
            return null;
        }

        return _byCode.TryGetValue(wireCode, out var entry) ? entry : null;
    }

    /// <inheritdoc />
    public override string ToString() => Code;

    private static ReadOnlyDictionary<string, ErrorCode> BuildIndex()
    {
        var dict = new Dictionary<string, ErrorCode>(_all.Count, System.StringComparer.Ordinal);
        foreach (var entry in _all)
        {
            dict[entry.Code] = entry;
        }

        return new ReadOnlyDictionary<string, ErrorCode>(dict);
    }
}
